zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationCatchEventBehaviour.java

```diff
@@ -125,6 +144,103 @@ public void handleCatchEvents(
         targetProcessDefinition, sourceElementIdToTargetElementId, signalSubscriptionsToMigrate);
   }
 
+  private void handleCompensationCatchEvents(
+      final DeployedProcess targetProcessDefinition,
+      final DeployedProcess sourceProcessDefinition,
+      final Map<String, String> sourceElementIdToTargetElementId,
+      final BpmnElementContextImpl context) {
+    final List<CompensationSubscription> compensationSubscriptionsToMigrate =
+        unsubscribeFromCompensationEvents(
+            sourceProcessDefinition, sourceElementIdToTargetElementId, context);
+    // DO NOT SUBSCRIBE TO UNMAPPED COMPENSATION EVENTS AS THEY ARE CREATED ON ELEMENT COMPLETION
+    migrateCompensationEvents(
+        targetProcessDefinition,
+        sourceElementIdToTargetElementId,
+        compensationSubscriptionsToMigrate);
+  }
+
+  private void migrateCompensationEvents(
+      final DeployedProcess targetProcessDefinition,
+      final Map<String, String> sourceElementIdToTargetElementId,
+      final List<CompensationSubscription> compensationSubscriptionsToMigrate) {
+    compensationSubscriptionsToMigrate.forEach(
+        subscription -> {
+          final var sourceActivityId = subscription.getRecord().getCompensableActivityId();
+          final var targetActivityId = sourceElementIdToTargetElementId.get(sourceActivityId);
+
+          final var compensationSubscriptionRecord = subscription.getRecord();
+          final var recordCopy = new CompensationSubscriptionRecord();
+          recordCopy.wrap(compensationSubscriptionRecord);
+          recordCopy.setProcessDefinitionKey(targetProcessDefinition.getKey());
+          recordCopy.setCompensableActivityId(targetActivityId);
```

Comment:
❓ I think we are missing some fields to set here, like the `processInstanceKey` or `compensationHandlerId` or `compensableActivityInstanceKey`

Reply:
`processInstanceKey` and `compensableActivityInstanceKey` should stay the same since they are related to the process **instance** (not the process definition) and process instance related data does not change during the migration. We only update process definition related data (ids, definition key etc.). However, `compensationHandlerId` is part of the definition and needs to be updated. I already send the change for that one.

Please let me know if you need more clarification around process instance migration internals.